Fix crash when connection closes during handshake

oderwat opened the pull request:
Fixes #223

FULL DISCLOSURE: Beware, I am an Erlang beginner that was using Claude AI (with MCPs) to create this PR (in 30 Minutes). I think I do understand what is being done, but I would not be able to write that easily from scratch.

- Replace direct pattern matching with case statements in handshake/8 and auth_finish_or_switch/6 to handle {error, Reason} from recv_packet
- Return proper #error{} records instead of crashing with unmatched pattern
- Add error codes -3 and -4 for handshake and authentication failures respectively
- Add test cases to verify graceful error handling

This ensures mysql-otp returns proper error results when MySQL connections are closed during handshake due to authentication issues, network problems, or server configuration issues, instead of crashing the application.

oderwat committed:
Fix crash when connection closes during handshake

Fixes #223

- Replace direct pattern matching with case statements in handshake/8
  and auth_finish_or_switch/6 to handle {error, Reason} from recv_packet
- Return proper #error{} records instead of crashing with unmatched pattern
- Add error codes -3 and -4 for handshake and authentication failures respectively
- Add test cases to verify graceful error handling

This ensures mysql-otp returns proper error results when MySQL connections
are closed during handshake due to authentication issues, network problems,
or server configuration issues, instead of crashing the application.

juhlig commented:
No offense, but I'm not willing to just accept AI-generated stuff. The nested case looks fishy, and the test suite is failing, too 🫤 I'll take a look on monday, on vacation right now.

zuiderkwast commented:
Yeah I saw the added nesting too. I hope we can break it out to a new function to avoid added nesting.

By checking the "hide whitespace changes" checkbox in the diff, it's clear that it's just indentation.

Regarding AI generated content, IMO it can be reviewed and accepted as any other contribution. The main thing to watch out for is if the AI reproduces copyrighted matierial under an incompatible license. There no risk here though where we're just adding a check in existing code.

Maria-12648430 reviewed (changes requested):
> Beware, I am an Erlang beginner

Excellent learning ground then :grin:

oderwat commented:
@Maria-12648430 [act](https://github.com/nektos/act) is a test runner that runs your original GitHub workflows locally (using Docker).

> Excellent learning ground then 😁

I didn't originally plan to learn Erlang when I started exploring Gleam, but it's been incredibly helpful for working around Gleam's incomplete features and type system limitations.

Looking back, I really wish I had discovered Erlang in 1998. Instead, I was just starting to build a PHP framework at the time - which, to be fair, paid a great share of my bills for the past 25+ years!

I'm pretty busy right now, but I'm happy to look into updating the PR if no one else gets to it first.

oderwat committed:
Refactoring after review requests.

This was done by AI, set to the PR conversation.

Please have a quick look, if that goes anywhere.

juhlig committed:
Straighten implementation and fix tests

oderwat commented on test/mysql_protocol_tests.erl:
```diff
@@ -135,6 +135,90 @@ error_as_initial_packet_test() ->
+    
+    %% Expect recv for handshake packet, but no recv for auth confirm packet
+    %% This will cause the auth_finish_or_switch to fail
+    ExpectedCommunication = [
+        {recv, HandshakePacket},
+        {send, ignore_this_send}  % The handshake response will be sent
+        %% No recv for auth confirmation - this will cause the error
+    ],
+    
+    Sock = mock_tcp:create(ExpectedCommunication),
```
should be rebased on master now

Maria-12648430 commented on src/mysql_protocol.erl:
```diff
@@ -76,21 +76,40 @@
-    {ok, HandshakePacket, SeqNum1} = recv_packet(SockModule0, Socket0, SeqNum0),
+    case recv_packet(SockModule0, Socket0, SeqNum0) of
+        {ok, HandshakePacket, SeqNum1} ->
+            handshake_process_initial_packet(HandshakePacket, Host, Username, Password, 
+                                           Database, SockModule0, SSLOpts, Socket0, 
+                                           SeqNum1, SetFoundRows);
```
```suggestion
            handshake_process_initial_packet(HandshakePacket, Host, Username, Password, 
                                             Database, SockModule0, SSLOpts, Socket0, 
                                             SeqNum1, SetFoundRows);
```

Maria-12648430 reviewed (changes requested):
LGTM, except the small indendation issue.

Maria-12648430 committed:
Fix indendation

Maria-12648430 reviewed (approved):
Fixed the indendation myself :)

Maria-12648430 commented:
Merged. Thanks for your Pull Request :slightly_smiling_face:
